src/vocabulary/mod.rs

```diff
@@ -0,0 +1,308 @@
+use std::collections::HashMap;
+
+use tokenizers::normalizers::Sequence;
+use tokenizers::{FromPretrainedParameters, NormalizerWrapper, Tokenizer};
+
+use crate::prelude::*;
+use crate::VocabularyError;
+
+use locator::EosTokenLocator;
+use processor::TokenProcessor;
+
+mod locator;
+mod processor;
+
+/// Vocabulary of an LLM.
+///
+/// ## Examples
+///
+/// ```rust
+/// # use outlines_core::prelude::*;
+/// #
+/// let vocabulary = Vocabulary::new(None)
+///     .insert("blah", 0)
+///     .insert("1a", 1)
+///     .insert("2", 2)
+///     .insert("0", 3);
+/// ```
+#[derive(Clone, Debug, Default)]
+pub struct Vocabulary {
+    // TODO: Option is temp for back compatibility
+    eos_token_id: Option<TokenId>,
+    map: HashMap<Token, Vec<TokenId>>,
+}
+
+impl Vocabulary {
+    /// Creates an empty vocabulary.
+    pub fn new(eos_token_id: Option<TokenId>) -> Self {
+        Self {
+            eos_token_id,
+            map: HashMap::new(),
+        }
+    }
+
+    /// Creates the vocabulary of pre-trained model from Hugging Face Hub.
+    pub fn from_pretrained(
+        model: &str,
+        parameters: Option<FromPretrainedParameters>,
+    ) -> Result<Self, VocabularyError> {
+        let mut tokenizer =
+            Tokenizer::from_pretrained(model, parameters.clone()).map_err(|error| {
+                VocabularyError::UnableToCreateTokenizer {
+                    model: model.to_string(),
+                    source: error,
+                }
+            })?;
+        Self::filter_normalizers(&mut tokenizer);
```

Comment:
Reasoning here was the consistency between the tokenizers versions:
- one being passed further for eos token id lookup
- another being used to build the vocabulary and process tokens
 
Even though we're not messing _right now_ with eos tokens in token processors (and probably never will), it creates ambiguity and if something changes for whatever reason, it might be a hard to debug logical issue.

Reply:
All good then!